src/multivariate/dirichlet.jl

```diff
@@ -156,18 +156,61 @@ end
 
 function _rand!(rng::AbstractRNG,
                 d::Union{Dirichlet,DirichletCanon},
-                x::AbstractVector{<:Real})
-    for (i, αi) in zip(eachindex(x), d.alpha)
-        @inbounds x[i] = rand(rng, Gamma(αi))
+                x::AbstractVector{E}) where {E<:Real}
+    
+    if any(a -> a >= .5, d.alpha)
+        # 0.5 is a placeholder; optimal value unknown
+        # 1 is known to be too high.
+        for (i, αi) in zip(eachindex(x), d.alpha)
+            @inbounds x[i] = rand(rng, Gamma(αi))
+        end
+
+        return lmul!(inv(sum(x)), x)
+    else
+        # Sample in log-space to lower underflow risk
+        for (i, αi) in zip(eachindex(x), d.alpha)
+            @inbounds x[i] = _logrand(rng, Gamma(αi))
+        end
+
+        if all(isinf, x)
```

Comment:
This seems somewhat incorrect - you're only looking for `-Inf` but checking for `Inf` and `-Inf`. For `Inf` already a single value is problematic. An additional question is why `isinf` instead of `!isfinite` which would e.g. also guard against `NaN`.